content/browser/ml/webnn/dml/readback_heap.cc

```diff
@@ -0,0 +1,121 @@
+  for (auto& output_info : outputs_info) {
+    uint64_t byte_length = reinterpret_cast<const DML_BUFFER_TENSOR_DESC*>(
+                               output_info.outputTensorDESC.Desc)
+                               ->TotalTensorSizeInBytes;
+    MemoryInfo memory_info = {};
+    memory_info.byte_offset = aligned_offset;
+    memory_info.byte_length = byte_length;
+    outputs_info_map_[output_info.name] = memory_info;
+
+    // Only offset need to be algnement, the byte length keep original value.
+    aligned_offset += Align(byte_length, DML_MINIMUM_BUFFER_TENSOR_ALIGNMENT);
+  }
+  outputs_resource_size_ = aligned_offset;
+  outputs_shm_region_ =
+      base::ReadOnlySharedMemoryRegion::Create(outputs_resource_size_);
+
+  HRESULT hr = CreateReadbackResource(outputs_resource_size_);
+  if (FAILED(hr)) {
+    return hr;
+  }
+  return S_OK;
+}
+
+// Readback inference result from GPU that is stored in named_outputs.
+HRESULT ReadbackHeap::ReadbackResource(NamedOutputsPtr& named_outputs,
+                                       ID3D12Resource* src_resource) {
+  // TODO:: Don't need add barrier to reset source resource from COPY_SOURCE
+  // to UNORDERED_ACCESS?
+  // Copy buffer from GPU resource to CPU data.
+  execution_context_->CopyBufferRegion(readback_resource_.Get(), src_resource,
+                                       outputs_resource_size_,
+                                       D3D12_RESOURCE_STATE_COPY_SOURCE);
+
+  execution_context_->Flush();
+
+  D3D12_RANGE tensorBufferRange{0, outputs_resource_size_};
+  int8_t* readBackBuffer;
+  WEBNN_CHECK(readback_resource_->Map(
+      0, &tensorBufferRange, reinterpret_cast<void**>(&readBackBuffer)));
+
+  for (auto& [name, memory_info] : outputs_info_map_) {
+    auto mojo_memory_info = ml::webnn::mojom::MemoryInfo::New();
+    size_t byte_offset = memory_info.byte_offset;
+    size_t byte_length = memory_info.byte_length;
+    mojo_memory_info->byte_offset = byte_offset;
+    mojo_memory_info->byte_length = byte_length;
+    named_outputs->outputs[name] = std::move(mojo_memory_info);
+
+    std::vector<uint8_t> output_buffer(byte_length);
+    uint8_t* address =
+        outputs_shm_region_.mapping.GetMemoryAs<uint8_t>() + byte_offset;
+    memcpy(address, readBackBuffer + byte_offset, byte_length);
+  }
+  named_outputs->shared_memory = outputs_shm_region_.region.Duplicate();
+
+  readback_resource_->Unmap(0, nullptr);
+  return S_OK;
+}
+
+size_t ReadbackHeap::GetOutputsResourceSize() const {
+  return outputs_resource_size_;
+}
+
+ReadbackHeap::MemoryInfo::MemoryInfo() = default;
+ReadbackHeap::MemoryInfo::~MemoryInfo() = default;
+
+HRESULT ReadbackHeap::CreateReadbackResource(size_t byte_length) {
+  D3D12_HEAP_PROPERTIES heap_properties;
+  // TODO::Support Unified Memory Architecture (UMA) that don't need to copy
+  // anything there because GPU heaps are always mappable by CPU on unified,
+  // D3D12_HEAP_TYPE_CUSTOM specify the memory pool and CPU cache properties
```

Comment:
FYI, we don't need to change the heap type (from UPLOAD to CUSTOM) to avoid the extra copy on UMA. The use of D3D12_HEAP_TYPE_CUSTOM is to allow for memory reuse on UMA by not using heap types. These are distinct and unrelated UMA optimizations.

Reply:
Remove the D3D12_HEAP_TYPE_CUSTOM, thanks.